### ISSUE TITLE
Please add windows support for tmux

### ISSUE THREAD

**Issue by soroshsabz** [tmux/tmux#2575]
ITNOA

I think it is very good and useful for windows community to add support windows for tmux, and tmux can run directly on PowerShell in windows.

How we can help you for doing this?

**Comment by nicm** [comment]
This is unlikely to happen, the Windows terminal is just too different. Also see https://github.com/tmux/tmux/issues/2368

**Comment by nicm** [comment]
I suppose we could do it without ANSI escape sequences, but it would not be easy.

**Comment by krader1961** [comment]
Note that the Windows 10 console has support for ANSI escape sequences; although, it has to be enabled by calling SetConsoleMode() with the ENABLE_VIRTUAL_TERMINAL_PROCESSING flag. See https://docs.microsoft.com/en-us/windows/console/setconsolemode as well as numerous StackOverflow questions on this topic. The [Elvish shell](https://elv.sh) enables that mode to produce styled output without having to otherwise know anything about MS Windows. There are, of course, other quirks such as a different signal handling model that may be problematic for tmux.

In short, it's probably possible to make a native Windows 10 version of tmux, without littering the code with lots of `#if WIN32` directives, but it will have some limitations. Such as not being able to dynamically enable/disable tracing via SIGUSR2. It's also wholly impractical to support older versions of MS Windows.

**Comment by soroshsabz** [comment]
I think Windows 10 support is just enough. 😊

**Comment by skyline75489** [comment]
As @krader1961 pointed out, Windows console now supports VT sequences just as normal Linux terminals. And there's the brand new [Windows Terminal](https://github.com/microsoft/terminal) as a replacement to conhost.exe to improve the overall console experience on Windows. You can use a large amount of ANSI escape sequences and it would work as expected, if you set the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` flag.

The major thing stopping tmux from being available on Windows is that tmux depends on ncurses, which does not have Windows support, for now. @nicm I'd like to purpose reopening this issue.

**Comment by nicm** [comment]
Are you going to work on it? If so then feel free to ask me questions, I will reopen the issue when things are clearly moving.

**Comment by skyline75489** [comment]
I might give it a try to see if it's possible to create an MinGW build of tmux. I'm an outside contributor of Windows Terminal so I know it's possible to have tmux on Windows and I'd personally love to see it happen.

**Comment by nicm** [comment]
I expect the problems you will have will be with file descriptor passing, the other socket stuff should be possible to get working. Something like what we do for Cygwin where we just open the ttyname in the server might work, I don't know anything about Windows terminal APIs

**Comment by skyline75489** [comment]
OK I did some investigation and found that there's already a `tmux` package inside Cygwin and it works inside MSYS2. But outside MSYS2 it errors out:

```text
open terminal failed: not a terminal
```

That's good. At least most of the code compiles just fine with MINGW.

**Comment by skyline75489** [comment]
For anyone who try to work on this in the future, https://github.com/microsoft/terminal/issues/9332 will probably still be the huge blocker that I failed to concur. The lack of interoperation ability between file handle (including stdin/stdout handle) & socket handle dramatically complicates the porting of tmux on Windows.

**Comment by nicm** [comment]
You would need to bypass libevent and manage the tty descriptor yourself.

The easiest way to do this would be:

- Open a socketpair and pass one end to a thread that blocks on the tty descriptor.
- Create a libevent event to read from the other end of the socketpair.
- When data is readable, the thread should read it into an evbuffer and attach that to a queue, with appropriate locking.
- When that finishes, it writes a byte into the socketpair.
- This triggers the libevent read callback, which consumes all the evbuffers from the queue, again with appropriate locking.

And you would need to do the reverse for writing to the socket.

There may be a way to do it without a queue and lock, if libevent evbuffer read/write/drain is thread safe.

Of course this would mean two extra threads for every pane. If Windows provides a way to wait on multiple tty descriptors at the same time you could do it with one thread and just say which descriptor/pane it came from when you add it to the queue.

Doing this in a way that is concealed enough from tmux not to cause major disruption would need some care. I would make it all hidden behind an API in compat/ that fires callbacks similar to libevent and then it should only need a few wrapper functions or ifdefs in window.c.

**Comment by skyline75489** [comment]
Despite the fact that there’s no ‘socketpair’ on Windows(!), that’s basically what I try to accomplish.

I’m having a little trouble understanding how tty.c works. I know that the server controls all the drawing of clients by manipulating the tty handle. How does server handle the output from shell? Like bash/zsh etc. I know the shells are spawned using forkpty but I didn’t find the code that handle the output of the shells, and combine it with the tmux staus line, etc.

**Comment by nicm** [comment]
Well I assume you would need to do the same for the client tty fd also, this is a little more complicated because tmux does the read/write itself rather than using a bufferevent. But there is no reason it couldn't do the same for window ptys and then one API could do both.

**Comment by nicm** [comment]
You can use `evutil_socketpair` to get a socketpair from libevent on Windows. Look for it here: http://www.wangafu.net/~nickm/libevent-book/Ref5_evutil.html

**Comment by nicm** [comment]
The output from panes is handled in `window.c`, it is a lot simpler than tty.c because it uses a bufferevent so libevent does most of the work. Look at `window_pane_set_event`, `window_pane_read_callback` and `window_pane_error_callback` functions.

There is no reason `window.c` could not do the work itself with `evbuffer_read` in the same way as `tty.c` - it just doesn't need to.

So if you provided an API with functions like the libevent ones used by `tty.c` (add event, remove event, read data once cb fired, write), both files could use it.

**Comment by nicm** [comment]
Basically we would need something like this:
~~~~
typedef void *tty_descriptor;
tty_descriptor *register_tty_descriptor(int td, event_callback_fn read_cb, event_callback_fn write_cb, void *arg);
void deregister_tty_descriptor(struct tty_descriptor *td);
struct evbuffer *read_tty_descriptor(tty_descriptor *td); // return read bytes (inside read_cb)
void drain_tty_descriptor(tty_descriptor *td, size_t used); // remove consumed bytes
ssize_t write_tty_descriptor(tty_descriptor *td, struct evbuffer *evb); // copy evb and queue for write
~~~~
Then for most platforms this could be backed by libevent and on Windows it could deliver to a helper thread.

You can add an `osdep_event_init` like on other platforms to call `evthread_use_windows_threads` or any other initialization.

It would be fine if write always returned success - the flow control wouldn't work but we could live without that too.

**Comment by skyline75489** [comment]
Using thread to read/write the pty handle is also the first idea in my mind. Funny that I thought there's no way you would want introduce extra threads, let along extra threads per pane since right now libevent use no thread at all. So I stop exploring the this option. Now that you're OK with it, I'll give a try when I have the time.

**Comment by skyline75489** [comment]
Using thread to read/write the pty handle is also the first idea in my mind. Funny that I thought there's no way you would want introduce extra threads, let along extra threads per pane since right now libevent use no thread at all. So I stop exploring the this option. Now that you're OK with it, I'll give a try when I have the time.

**Comment by nicm** [comment]
It would be better to use 1 thread for all panes, but I don't really mind how it works on Windows so long as it can be self-contained and doesn't make a mess of the rest of the codebase.

**Comment by skyline75489** [comment]
There's this line that confuses me so much:

https://github.com/tmux/tmux/blob/de4ac37baa2d32c3a3597be39eeac953c6159f77/tty.c#L268

So here `c->fd` is the STDIN sent from client using `MSG_IDENTIFY_STDIN`, and `tty->event_out` is the output buffer that buffers things before sending them. I don't get how the `event_set` works. My understanding is that `event_set` monitors any `EV_WRITE` on `c->fd`. But does anyone else write to `c->fd` except tmux server itself ? If no one writes to STDIN, the event will not be fired. If tmux server itself writes to STDIN, will it just go recursively and trap itself?..

But I guess I can just skip the buffer & flow control things on Windows at first.

**Comment by nicm** [comment]
`EV_WRITE` does not tell you when *someone else* writes to the event, it tells you when the event is *ready to write*. So when the callback is fired, tmux knows that it can write data to the file descriptor.

Note that `event_set` just initializes the event, it has be added with `event_add` before it will do anything.

**Comment by nicm** [comment]
Basically it is the difference between `POLLIN` and `POLLOUT` with `poll` or `readfds` and `writefds` for `select`.

**Comment by skyline75489** [comment]
Oh, thanks for the explanation. Now I see how this works. 

So `tty.c` handles most logic about reading from & writing to the TTY sent from client. And `windows.c` handles most logic communicating with the PTY created by `forkpty`, through which shells like `bash` is spawned. Is this correct?

Pardon me for not really knowing Linux tty that well. I failed to find the connection point where the output of PTY is sent to the client. Or is there anything like it? 

On Windows there's PTY like API called `ConPTY`. But for STDIN/STDOUT there's no such file descriptor as tty so I'll need do extra work for that.

**Comment by nicm** [comment]
Yes that is right.

There is no need to send output to the client because tmux owns a file
descriptor that connects directly to the terminal device (the stdin that
the client sent it), so when tmux writes to it it goes there without
needing to go to the client first.

**Comment by rashil2000** [comment]
Hi @nicm, sorry for hijacking the thread here, but before a truly-native Windows port of tmux arrives, is there a temporary fix for tmux so that it can work in Cygwin/Msys2 environment, but on terminals other than mintty? Currently, using Cygwin/Msys2 environment on other terminals, tmux just gives the error that @skyline75489 mentioned. 

[This mailing list thread](http://cygwin.1069669.n5.nabble.com/quot-tmux-open-terminal-failed-not-a-terminal-quot-in-terminal-emulators-other-than-mintty-td150569.html) goes on in some detail as to what exactly might be happening.

**Comment by soroshsabz** [comment]
@skyline75489 Is there any plan to make PR for this issue? or instead of that make some detail explanation / design to How to add windows native compatibility?

thanks

**Comment by skyline75489** [comment]
@soroshsabz I’ve done some PoC kind of work privately and found it almost impossible to add Windows support without a ton of changes in tmux itself. So I gave up on the idea. See https://github.com/tmux/tmux/pull/2599#issuecomment-796462166 for details.

**Comment by evandrocoan** [comment]
I checkout this master branch built everything successfully, but when running it, I was getting `error creating /tmp/tmux-197609 (Permission denied)` when trying to opening tmux. I deleted everything in my `/tmp/` directory and the problem persisted.

The problem was happening because the msys2 package for tmux does some patches on the code, to fix it, I need to first apply those patches, only then to install the package.
```
git clone https://github.com/msys2/MSYS2-packages/
cd MSYS2-packages/tmux

vim PKGBUILD
# and replace `_base_ver=3.2` by `master`
# and put on the first line of `sha256sums=('29....'` 
# the sha256 of the file https://github.com/tmux/tmux/archive/master.tar.gz

makepkg
pacman -U tmux-master-1-x86_64.pkg.tar.zst
```
